[3기 이한솔]TodoList with Ajax

catsbi opened the pull request:
Footer부분을 제외한 기본적인 기능구현하여 PR합니다 .
프론트엔드 상태관리개념과, Observer Pattern의 기능구현을 적용하려니 익숙하지않아, 시간이 너무지체되어 
미완성이라도 올렸습니다.

상위컴포넌트이자 ConcreteSubject인 TodoApp에서 각 이벤트에 대응하는 로직을 넣어주려하다가 너무 양이 많아
서비스클래스(UserService)를 만들어서 내려주는 방식으로 했습니다. Dependancy Injection을 적용하고싶어 알아보고 있습니다.. 
그게아니라면 해당 서비스로직들을 스태틱하게 만들어주는것과 고려중입니다.

모든 동작에 새로 렌더링을 통해 그려주는 방식으로 구현을 하려하니, 몇 애니메이션 효과들을 볼 수 없는데, 이런부분은 어떻게 처리해야할지 고민입니다. 

기능 구현을 하는 하나하나가 어렵다기보다는 이 로직이랑 이런 부분은 어디에 위치해야하고, 데이터들은 어디서 관리해야하고
그런 계층을 나누고 데이터들의 위치를 정하는것이 너무 헷갈리고 지금도 계속 혼동되어 수정과 롤백이 반복되고 있네요..

- 구현된 기능::요구사항
1. User 추가하기
2. User의 todolist 불러오기
3. User삭제하기 - 기능은 완료되었으나 UI단에 어디에 붙혀야 하는지 버튼이나 디자인이 없어서 보류
4. todoItem 추가하기
5. todoItem 불러오기
6. todoItem complete하기
7. todoItem 삭제하기

- 구현된 기능::심화 요구사항
1. fetch api를 사용하는 부분을 async await을 사용하여 리팩토링
2. 우선순위 뱃지 - span으로 변경은 아직 보류
3. ES6 impot & export를 이용

catsbi committed:
2주차 미션, Footer 부분제외, 기본적인 기능 구현

JunilHwang commented on index.html:
```diff
@@ -3,6 +3,9 @@
   <meta charset="utf-8" />
+  <script>
+
+  </script>
```
실수로 지우지 않으신 것 같아요!

JunilHwang commented on index.html:
```diff
@@ -122,5 +110,6 @@ <h1 id="user-title" data-username="eastjun">
   </section>
 </div>
 <script src="js/index.js"></script>
```
마찬가지로 지우지 않으신 것 같습니다!

JunilHwang commented on js/App.js:
```diff
@@ -0,0 +1,17 @@
+import { TodoApp } from "./components/TodoApp.js";
+import { TodoRepository } from "./repository/TodoRepository.js";
+
+window.addEventListener('DOMContentLoaded', async (event) => {
+    /*const todoApp = new TodoApp();
+    todoApp.addObservers(
+        new TodoInput(todoApp),
+        new TodoList(todoApp),
+        new TodoCount(todoApp)
+    );
+    todoApp.render();*/
+    let todoRepository = new TodoRepository();
+    let findUsers = await todoRepository.findUsers();
+    debugger
+
+
+});
```
App.js가 작성은 되었는데 html에서도 그렇고 js에서도 그렇고 가져오는 코드가 없네요!

catsbi replied:
전체적으로 코드를 하도 수정&재정의하면서 익숙치않은 방식의 설계로 하다보니 사용하지않는 잔여 코드들이 너무 많이 남아있네요. 오늘 정리할 예정입니다 ㅠㅠ

JunilHwang commented on js/components/TodoApp.js:
```diff
@@ -0,0 +1,70 @@
+        let userList = new UserList(_.userListTarget, this);
+        let todoInput = new TodoInput(_.todoInputTarget, this);
+        let todoList = new TodoList(_.todoListTarget, this);
+        let todoFooter = new TodoFooter(_.todoFooterTarget, this);
```
여기서 변수는 const로 하는게 맞는 것 같아요!

JunilHwang commented on js/components/TodoApp.js:
```diff
@@ -0,0 +1,70 @@
+            todoList,
+            todoFooter
+        );
+        this.notify();
```
컴포넌트가 재사용되고 있진 않기 때문에 그냥 addObservers인자에서 바로 생성해주는 것도 좋았을 것 같습니다!

JunilHwang commented on js/components/TodoApp.js:
```diff
@@ -0,0 +1,70 @@
+
+    getService = () =>{
+        return this.#userService;
+    }
```
```suggestion
    getRepository = () => this.#todoRepository;
    getService = () => this.#userService;
```
이렇게 표현할 수 있습니다!

JunilHwang commented on js/observer/Observer.js:
```diff
@@ -0,0 +1,29 @@
+
+    render(subject) {
+    }
+};
```
observer의 역할과 component의 역할이 같이 존재하는 것 같습니다!
observer를 완전히 분리한 다음에
component가 observer를 상속받고,
마찬가지로 현재 observer를 상속받고 있는 것들이 component를 상속받도록 하면 어떨까요?

JunilHwang commented on js/observer/Observer.js:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    render(subject) {
+    }
```
지금 template 코드가 완전히 분리되어 있는 상태입니다.
template이 완전히 분리되어 있을 경우 오히려 유지보수할 때 여러 파일을 왔다 갔다 해야 해서 불편할 것 같아요! 특히 template 코드가 한 파일 내에 같이 있다면 더 불편할 것 같습니다.

차라리 component 내부에 template 이라는 메소드가 존재하고
render를 실행 시에 
template 메소드에 정의한걸 사용하는 방식은 어떨까요?

```js
render () {
    this._target.innerHTML = this.template();
}
```
이런식으로여!

JunilHwang commented on js/observer/Subject.js:
```diff
@@ -0,0 +1,28 @@
+        this.#observers = new Set(observers);
+    }
+
+    addObservers(...observers) {
```
addObserver는 외부에서 실행하지 않는 메소드 같습니다.
private method로 접근제한을 하면 어떨까요?

JunilHwang commented on js/repository/Repository.js:
```diff
@@ -0,0 +1,28 @@
+    #url
+
+    constructor(url) {
+        this.#options = {
+            method: "GET",
+            dataType: "JSON",
+            headers: {
+                'Content-Type': 'application/json'
+            }
+        };
```
this.#options도 인자로 받아오고 더 추상화하여 관리하기가 좋을 것 같아요!
```js
        this.#options = {
            method: "GET",
            dataType: "JSON",
            headers: {
                'Content-Type': 'application/json'
            },
            ...options // 이 부분만 추가하면 된답니다!
        };
```

JunilHwang commented on js/repository/Repository.js:
```diff
@@ -0,0 +1,28 @@
+        this.#options = { ...this.#options, ...options };
+    }
+
+    fetch(method = "GET", path, body = null, before, after) {
```
body에 인자가 없을 경우 null이 들어가도록 되어 있습니다.
이럴 때 body 뒤에 있는 before, after 인자가  문제가 돼요!
before, after를 받아오기 위해선 결국 body를 null 형태로라도 인자로 보내야하니까요.

차라리 body를 맨 뒤에 놓거나 혹은 before와 aftter에도 기본값을 지정하는게 좋을 것 같습니다.

JunilHwang commented on js/repository/Repository.js:
```diff
@@ -0,0 +1,28 @@
+    fetch(method = "GET", path, body = null, before, after) {
+        this.#changeOptions({ method, body:body?JSON.stringify(body):null });
+        if(before)before();
+        let response = fetch(this.#url + path, this.#options).then(data=>data.json());
+        if(after)after();
```
이 부분은 순서가 보장되지 않는 것 같아요~
after가 완전히 응답을 받아온 뒤에 실행될 코드라면 좋겠지만, 제 생각에 이 부분은 로딩 떄문에 만들어주신 것 같은데 이러면 로딩으로서의 역할을 제대로 수행할 수 없을 것 같습니다.
response.then 부분에서 after를 수행해야 하는게 아닐까요?

JunilHwang commented on js/repository/Repository.js:
```diff
@@ -0,0 +1,28 @@
+export const Repository = class {
```
그리고 보통 Repository라고 하면 DB나 저장소를 의미하기 때문에
API 통신을 하는 코드에 어울리는 이름은 아닌 것 같아요!

JunilHwang commented on js/repository/TodoRepository.js:
```diff
@@ -0,0 +1,75 @@
+
+    return defaultOptions;
+}
+
```
앞서 언급했던 것 처럼 body 뒤에 인자를 모두 null로 넘기고 있습니다. after와 before에도 기본값을 만들어주시는게 좋을 것 같아요!
```js
before = () => {},
after = () => {}
```
이렇게 단순하게 아무 역할도 하지 않는 함수로 만들면 될 것 같습니다!

JunilHwang commented on js/repository/TodoRepository.js:
```diff
@@ -0,0 +1,75 @@
+    }
+
+     findUsers() {
+        return this.fetch("GET", "/users", null, null, null);
```
GET, DELEET에 대한 전용 fetch 요청과
POST, PUT 등에 대한 전용 fetch 요청으로 구분하면 어떨까요?
그리고 fetch 메소드에서 인자로 http method를 선택하여 보내는 요청보단
this.get, this.delete, this.put, this.post
이렇게 명확하게 어떤 요청을 보낼지 메소드로 정의하면 좋을 것 같습니다!

JunilHwang commented on js/repository/TodoRepository.js:
```diff
@@ -0,0 +1,75 @@
+
+     saveUser(name) {
+        try {
+            return this.fetch("POST", "/users", { name }, null, null);
```
"POST" 라는 문자열보단
HttpMethod.POST
처럼 constant의 변수로 보내는건 어떨까요?

JunilHwang commented on js/service/UserService.js:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    addItem = async (contents) => {
+        let todoItem = await this.#todoRepository.saveTodoItem(this.#selectedUser._id, contents);
+        this.#selectedUser.todoList.push(todoItem);
+        this.#subject.notify();
+    }
+    deleteItem = async (itemId) => {
+        let todoList = this.#selectedUser.todoList;
+        let findIndex = todoList.findIndex(item => item._id === itemId);
```
변수는 일단 const로 정의해준다고 생각해주세요! let의 말 그대로 "변수"를 선언할 때 사용합니다. 하지만 생각보다 선언한 변수의 값이 변경되는 경우는 많지 않아요!

JunilHwang commented on js/service/UserService.js:
```diff
@@ -0,0 +1,115 @@
+ */
+
+
+export const UserService = class {
```
Service, Repository 등의 코드를 singleton으로 디자인해보는건 어떨까요?

catsbi replied:
소스코드의 최상위 컴포넌트(Subject) 혹은 App부분에서 선언된 클래스를 선언해서 주입하는방식과 싱글톤방식중 어느게 더 나을까요? 

스프링을 하면서 하던 개념으로만 접근을해서 싱글톤보단 초창기 생성후 공유하는방식을 사용했는데, 
싱글톤도 고려사항중 하나였거든요. 

그러다가 그냥 익숙하게 하던대로 해보자라고해서 Subject부분에서 생성후 참조를 넘겨주는 방식으로했습니다. 
어느 방법을 선택해야하는게 더 좋은가에 대한 고민을 항상하는데 부족함이 너무 많네요 ㅠ

JunilHwang replied:
음 제 생각엔 어느 한가지가 정답이라기 보단 상황에 맞게 적절하게 사용하는게 중요한 것 같아요!
사실 스프링 방식처럼 하기 위해선 아예 스프링 컨테이너처럼 빈으로 등록하면 싱글톤으로 사용할 수 있도록 만드는게 베스트인데.. 그렇게까지 하기에는 또 쉽지 않아서요 ㅠㅠ 저의 경우 각 객체마다 고유의 멤버 변수가 필요하다면 그냥 객체로 사용하고
싱글톤 형태로 사용해야 한다면 그냥 Object로 만들어서 사용하고 있어요!

JunilHwang commented on js/service/UserService.js:
```diff
@@ -0,0 +1,115 @@
+        return todoList.find(item => item._id === itemId);
+    };
+
+    #decodePriority(priority) {
+        return priority === "0"
+            ? "NONE" : priority === "1"
+                ? "FIRST" : "SECOND"
+    }
```
priority도 상수로 관리하면 좋을 것같아요!

JunilHwang commented on js/components/TodoApp.js:
```diff
@@ -0,0 +1,70 @@
+        super();
+        this.#todoRepository = new TodoRepository();
+        this.#userService = new UserService(this.#todoRepository, this);
+        this.init({
+            userTitleTarget,
+            userListTarget,
+            todoInputTarget,
+            todoListTarget,
+            todoFooterTarget,
```
잘 생각해보면 todo의 변화가 user에게 전파되고 있진 않습니다.
그래서 서로 다른 observer를 가지는게 좋지않을까요?
지금은 todo에서 변화가 발생하면 user도 변하는 형태인데
현재 요구사항대로라면 user가 변화면 todo가 변할순 있어도 todo가 변했을 때 user가 변화하면 안 되는 상태입니다!

catsbi replied:
User와 TodoItem을 분리해서 따로 관리하자는 말씀이신가요? 그 부분은 고민하다가 시간이 부족해 패스했는데, 조언해주신거보면 맞는 말씀이신거같네요.

JunilHwang replied:
넵 그래서 이렇게 최상위에서 옵저버를 추가하는 방법이 있을 수도 있고,
혹은 각각의 컴포넌트에서 변화를 알리고 싶은 서브젝트를 추가하는 방법도 있습니다!

혹은 `Object.defineProeprty`를 이용해서 아예 자동화를 시킬수도 있어요!

예를들어
어떤 상태 A가 있고, 어떤 컴포넌트 B가 있습니다.

- A에 Object.defineProperty를 이용하여 A의 값이 호출(get)되었을 때 해당 값을 호출한 컴포넌트를 옵저버로 등록
- 그래서 A의 값이 변경될 때(set) A의 변화를 옵저버들에게 알려줌. 옵저버들은 변화된 값에 따라서 render를 실행
- 이 때 렌더링 코드가 최초로 한 번 호출되는 시점에만 옵저버로 등록

대신 이런식으로 디자인할 경우 조금 헷갈릴 수 있는게 컴포넌트가 서브젝트가 되기도 하고 옵저버가 되기도 합니다.
상위 컴포넌트의 변화를 자식에게 알려야할 수 있거든요!

JunilHwang reviewed (approved):
안녕하세요 한솔님! 몇 가지 리뷰 남겼습니다 ㅎㅎ 
전체적으로 잘 설계해주신 것 같아요! 지난 스텝보다 훨씬 코드가 좋아진게 느껴집니다.

제 리뷰도 잘 부탁드려요!!

catsbi committed:
1. 불필요 코드 제거
2. API호출 클래스 이름 변경(HttpClient)
3. TodoHttpClient 어댑터 추가
4. templates제거 및 컴포넌트에서 담당하도록 수정
5. HttpMethod및 몇몇 변수값들 상수화
6. todoList 하단부분 기능구현

jho2301 commented on js/components/TodoApp.js:
```diff
@@ -0,0 +1,58 @@
+    userListTarget: document.querySelector("#user-list"),
+    todoInputTarget: document.querySelector(".input-container"),
+    todoListTarget: document.querySelector(".todo-list"),
+    todoFooterTarget: document.querySelector(".count-container")
```
성능적으로 차이가 있다고 말씀해주셔서 찾아봤는데,

```
 Function               | Live? | Type           | Time Complexity
querySelector          |   N   | Element        |  O(n)
querySelectorAll       |   N   | NodeList       |  O(n)
getElementById         |   Y   | Element        |  O(1)
getElementsByClassName |   Y   | HTMLCollection |  O(1)
getElementsByTagName   |   Y   | HTMLCollection |  O(1)
getElementsByName      |   Y   | NodeList       |  O(1)

```

라고 하네요.

https://stackoverflow.com/questions/14377590/queryselector-and-queryselectorall-vs-getelementsbyclassname-and-getelementbyid#:~:text=querySelector%20and%20querySelectorAll%20are%20a,browsers%20you%20need%20to%20support.

https://hashcode.co.kr/questions/5692/%EA%B0%95%EC%9D%98-4-11-queryselector%EC%97%90-%EC%84%B1%EB%8A%A5%EB%AC%B8%EC%A0%9C%EC%97%90-%EB%8C%80%ED%95%B4-%EC%A7%88%EB%AC%B8-%EB%93%9C%EB%A6%BD%EB%8B%88%EB%8B%A4

catsbi replied:
막연이 더 빠르다정도만 알고있었는데, 링크주신글을 보니 저도 좀 더 자세히 알꺼같네요 감사합니다 ㅎ

jho2301 commented on js/data/constant.js:
```diff
@@ -0,0 +1,41 @@
+    static FIRST = "FIRST"
+    static SECOND = "SECOND"
+    static NONE = "NONE";
+}
```
class로 객체를 만들지 않으니, 객체를 export하면 어떨까요?

jho2301 commented on js/repository/HttpClient.js:
```diff
@@ -0,0 +1,52 @@
+        return fetch(this.#url + path, this.#options).then(data => data.json());
+    }
+
+    _excute(path) {
```
execute 함수가 중복인 것 같아요. private이라 그렇다면 public으로 바꾸는 게 어떨까요?

jho2301 commented on js/utils/functionalJS.js:
```diff
@@ -0,0 +1,122 @@
+        res.push(i);
+    }
+    return res;
+};
```
이게 말씀해주신 내용이네요. 함수형프로그래밍 학습에 대한 필요성을 느낍니다 ㅜㅜ.

jho2301 reviewed (commented):
안녕하세요 한솔님!
코드리뷰 남겼습니다!
리뷰 잘 부탁드립니다
